React/Testing best practices.md

```diff
@@ -0,0 +1,610 @@
+```tsx
+const UserPage: NextPage = () => {
+  const { data, error } = useSWR<Array<User>, IResponseError>(
+    USER_KEY,
+    fetchUsers
+  );
+
+  if (error) {
+    return <ErrorPage />;
+  }
+
+  if (!data) {
+    return <LoadingPage />;
+  }
+
+  return <UserTemplate userList={data} />;
+};
+
+export default UserPage;
+```
+
+It this page example we should test all three states of the page component. To do that we should mock `<ErrorPage>`, `<LoadingPage>` and `<UserTemplate>` and check if it's rendered based on fetcher response.
+
+Test example:
+
+```tsx
+jest.mock("components/error-page");
+jest.mock("components/loading-page");
+jest.mock("components/user-template");
+jest.mock("fetchers/users");
+
+(ErrorPage as jest.Mock).mockReturnValue(
+  <div data-testid="error-page-testid" />
+);
+(LoadingPage as jest.Mock).mockReturnValue(
+  <div data-testid="loading-page-testid" />
+);
+(UserTemplate as jest.Mock).mockReturnValue(
+  <div data-testid="user-template-page-testid" />
+);
+
+describe("User Page", () => {
+  it("Is error page rendered", () => {
+    (fetchUsers as jest.Mock).mockResolvedValue(new Error("Error occurred!"));
+
+    const { findByTestId } = render(<UserPage />);
+
+    expect(findByTestId("error-page-testid")).toBeDefined();
+  });
+
+  it("Is loading page rendered", () => {
+    (fetchUsers as jest.Mock).mockResolvedValue(null);
+
+    const { findByTestId } = render(<UserPage />);
+
+    expect(findByTestId("loading-page-testid")).toBeDefined();
+  });
+
+  it("Is loading page rendered", () => {
+    (fetchUsers as jest.Mock).mockResolvedValue([{ username: "test user" }]);
+
+    const { findByTestId } = render(<UserPage />);
+
+    expect(findByTestId("user-template-page-testid")).toBeDefined();
+  });
+});
+```
+
+Because of testing `useSWR` behavior which will re-render DOM after fetcher promise is resolved we need to use `findByTestId` method that returns a promise that will resolve when element is added to DOM.
+
+### SWR testing
+
+Testing swr had few known issues. One is cache persistent across tests. To fix that we need to add this:
+
+```tsx
+afterEach(async () => {
+  await waitFor(() => cache.clear());
+});
+```
+
+Another issue is deduping. We can fix that by adding `SWRConfig` provider to our custom render.
+
+```tsx
+const AllProviders = ({ children }) => (
+  <SWRConfig value={{ dedupingInterval: 0 }}>
+    <ChakraProvider theme={theme}>{children}</ChakraProvider>
+  </SWRConfig>
+);
+```
+
+More can be found on [this issue](https://github.com/vercel/swr/issues/781).
+[Additional info](https://medium.com/frontend-digest/using-testing-libary-with-useswr-f595919de2fd).
+
+## Fetchers
+
+Fetcher tests should be located in `/fetchers/{{ fetcher name }}` folder next to the fetcher that is tested.
+
+### Using datx:
+
+When testing fetchers that use Datx, we create mocked store Datx store and instead of making API calls with `request` we need to mock `request` and test just test functionality of the fetcher.
```

Comment:
```suggestion
When testing fetchers that use Datx, we create mocked store Datx store and instead of making API calls with `request` we need to mock `request` and test the functionality of the fetcher.
```